docs/node/interceptors.md

```diff
@@ -0,0 +1,171 @@
+
+```ts
+import { authenticate } from "./authenticate.js"; // This can come from an auth library like passport.js
+
+const authenticator: Interceptor = (next) => async (req) => {
+  // `authenticate` takes the authorization header value
+  // and returns the user that represents the token.
+  const user = authenticate(req.header.get("Authorization"));
+  if (user === undefined) {
+    throw new ConnectError("User not authenticated", Code.Unauthenticated);
+  }
+  return await next(req);
+};
+```
+
+But what if we need the user info in one of our rpc implementations? One way is to parse the header again:
+
+```ts
+import { ConnectRouter } from "@connectrpc/connect";
+import { ElizaService } from "./gen/eliza_connect";
+import { authenticate } from "authenticate.js";
+
+export default (router: ConnectRouter) =>
+  // registers connectrpc.eliza.v1.ElizaService
+  router.service(ElizaService, {
+    // implements rpc Say
+    async say(req, context) {
+      const user = authenticate(context.requestHeader.get("Authorization"))!;
+      return {
+        sentence: `Hey ${user.name}! You said: ${req.sentence}`,
+      };
+    },
+  });
+```
+
+But this means authentication happens twice, once in the Interceptor and second in our handler. This is where context values become useful, we
+can modify the interceptor to pass the user information:
+
+```ts
+import { authenticate } from "./authenticate.js"; // This can come from an auth library like passport.js
+import { kUser } from "user-context.js";
+
+const authenticator: Interceptor = (next) => async (req) => {
+  // `authenticate` takes the authorization header value
+  // and returns the user that represents the token.
+  const user = authenticate(req.header.get("Authorization"));
+  if (user === undefined) {
+    throw new ConnectError("User not authenticated", Code.Unauthenticated);
+  }
+  // Add the user to the request context.
+  req.contextValues.set(kUser, user);
+  return await next(req);
+};
+```
+
+And then in our handler we can use it:
+
+```ts
+import { ConnectRouter } from "@connectrpc/connect";
+import { ElizaService } from "./gen/eliza_connect";
+import { authenticate } from "authenticate.js";
+import { kUser } from "user-context.js";
+
+export default (router: ConnectRouter) =>
+  // registers connectrpc.eliza.v1.ElizaService
+  router.service(ElizaService, {
+    // implements rpc Say
+    async say(req, context) {
+      const user = values.get(kUser);
+      return {
+        sentence: `Hey ${user.name}! You said: ${req.sentence}`,
+      };
+    },
+  });
+```
+
+Here `kUser` is the key used to map the context values:
+
+```ts
+import { createContextKey } from "@connectrpc/connect";
+import type { User } from "./authenticate.js";
+
+const kUser = createContextKey<User>(
+  { name: "Anonymous" }, // Default value
+  {
+    description: "The current user", // Optional description, useful for debugging
+  },
+);
+
+export { kUser };
+```
```

Comment:
We can make this code snippet much more compact for moving it further up to the usage example:

```ts
import { createContextKey } from "@connectrpc/connect";

type User = { name: string };

const kUser = createContextKey<User>(
  { name: "Anonymous" }, // Default value
);
```

As mentioned above, context keys have some nice properties we should mention. We can use this short code snippet in the usage example, and maybe add a separate sub section for context keys that explains them in more depth, giving guidance on where to define them (in a separate file, or in the interceptor that provides them?), explaining the options (description).